File: e2e/node/tests/links.test.ts
import { describe, it, expect } from 'vitest';
import { scrape } from '@kreuzberg/kreuzcrawl';

describe('links', () => {
  it('links_anchor_fragment: Identifies fragment-only links as anchor type', async () => {
    const result = await scrape();
    expect(result.links[""].linkType).toContain("anchor");
  });

  it('links_base_tag: Resolves relative URLs using base tag href', async () => {
    const result = await scrape();
    expect(result.links.length).toBeGreaterThan(2);
    expect(result.links[""].url).toContain("example.com");
  });

  it('links_document_types: Detects PDF, DOCX, XLSX links as document type', async () => {
    const result = await scrape();
    expect(result.links[""].linkType).toContain("document");
  });

  it('links_empty_href: Handles empty href attributes without errors', async () => {
    const result = await scrape();
    expect(result.links.length).toBeGreaterThan(0);
    expect(result.links[""].url).toContain("/valid");
  });

  it('links_internal_external_classification: Correctly classifies internal vs external links by domain', async () => {
    const result = await scrape();
    expect(result.links.length).toBeGreaterThan(4);
    expect(result.links[""].linkType).toContain("internal");
    expect(result.links[""].linkType).toContain("external");
  });

  it('links_mailto_javascript_skip: Skips mailto:, javascript:, and tel: scheme links', async () => {
    const result = await scrape();
    expect(result.links.length).toBeGreaterThan(0);
    expect(result.links[""].url).not.toContain("mailto:");
  });

  it('links_protocol_relative: Handles protocol-relative URLs (//example.com) correctly', async () => {
    const result = await scrape();
    expect(result.links.length).toBeGreaterThan(1);
    expect(result.links[""].protocolRelative.length).toBeGreaterThan(0);
  });

  it('links_rel_attributes: Preserves rel=nofollow and rel=canonical attributes', async () => {
    const result = await scrape();
    expect(result.links.length).toBeGreaterThan(0);
  });

  it('links_relative_parent: Resolves ../ and ./ relative parent path links correctly', async () => {
    const result = await scrape();
    expect(result.links.length).toBeGreaterThan(3);
  });
});

File: e2e/rust/tests/links_test.rs
//! E2e tests for category: links

use kreuzcrawl::scrape;

#[test]
fn test_links_anchor_fragment() {
    // Identifies fragment-only links as anchor type
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.links.get("").map(|s| s.as_str()).link_type.contains(r#"anchor"#), "expected to contain: {}", r#"anchor"#);
}

#[test]
fn test_links_base_tag() {
    // Resolves relative URLs using base tag href
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.links.len() > 2_f64, "expected > 2");
    assert!(result.links.get("").map(|s| s.as_str()).url.contains(r#"example.com"#), "expected to contain: {}", r#"example.com"#);
}

#[test]
fn test_links_document_types() {
    // Detects PDF, DOCX, XLSX links as document type
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.links.get("").map(|s| s.as_str()).link_type.contains(r#"document"#), "expected to contain: {}", r#"document"#);
}

#[test]
fn test_links_empty_href() {
    // Handles empty href attributes without errors
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.links.len() > 0_f64, "expected > 0");
    assert!(result.links.get("").map(|s| s.as_str()).url.contains(r#"/valid"#), "expected to contain: {}", r#"/valid"#);
}

#[test]
fn test_links_internal_external_classification() {
    // Correctly classifies internal vs external links by domain
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.links.len() > 4_f64, "expected > 4");
    assert!(result.links.get("").map(|s| s.as_str()).link_type.contains(r#"internal"#), "expected to contain: {}", r#"internal"#);
    assert!(result.links.get("").map(|s| s.as_str()).link_type.contains(r#"external"#), "expected to contain: {}", r#"external"#);
}

#[test]
fn test_links_mailto_javascript_skip() {
    // Skips mailto:, javascript:, and tel: scheme links
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.links.len() > 0_f64, "expected > 0");
    assert!(!result.links.get("").map(|s| s.as_str()).url.contains(r#"mailto:"#), "expected NOT to contain: {}", r#"mailto:"#);
}

#[test]
fn test_links_protocol_relative() {
    // Handles protocol-relative URLs (//example.com) correctly
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.links.len() > 1_f64, "expected > 1");
    assert!(!result.links.get("").map(|s| s.as_str()).protocol_relative.is_empty(), "expected non-empty value");
}

#[test]
fn test_links_rel_attributes() {
    // Preserves rel=nofollow and rel=canonical attributes
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.links.len() > 0_f64, "expected > 0");
}

#[test]
fn test_links_relative_parent() {
    // Resolves ../ and ./ relative parent path links correctly
    let engine = None;
    let url = None;
    let result = scrape(engine, url).expect("should succeed");
    assert!(result.links.len() > 3_f64, "expected > 3");
}

